internal/consensus/state.go

```diff
@@ -801,12 +801,8 @@ func (cs *State) receiveRoutine(maxSteps int) {
 			cs.handleMsg(mi)
 
 		case mi = <-cs.internalMsgQueue:
-			err := cs.wal.WriteSync(mi) // NOTE: fsync
-			if err != nil {
-				panic(fmt.Sprintf(
-					"failed to write %v msg to consensus WAL due to %v; check your file system and restart the node",
-					mi, err,
-				))
+			if err := cs.wal.Write(mi); err != nil {
+				cs.Logger.Error("failed writing to WAL", "err", err)
```

Comment:
I think this is a pretty substantial change because it: 

- doesn't force an FSYNC, so we write to a WAL but don't force it to disk. 
- causes the node to continue if there's an error condition, in a situation where an error condition might mean that the node wouldn't recover consensus at the same point that it did when it crashed. 

My main concern (which I got from talking to Ethan during office hours a while ago) about changing WAL behavior is that it could make it possible for nodes to loose track of something it signed before it crashed, and the current WAL protects us from this.

Reply:
Yes, but writing is OK in this case. There is no need to flush until the `signVote` process.

Reply:
I think this is part of the reasons why we shouldn't have our own WAL implementation. The point of the WAL, as I understand it, is to add crash resistance in the case of "fast reboots' where the process stop/starts quickly in the middle of a consensus round. 

Every time we write, without flushing, we weaken this property, because more things can happen that we'll miss and have to be re-gossiped to us before the node can make progress, which probably means that a node that stops will miss out on the entire round. Having said this, I think the prospect of these kinds of "quick crashes" are relatively uncommon and the use case for the WAL (entirely) is a bit tangential, as most restarts will take longer and a node that restarts will (almost always) miss out on (at least one block.)

I'd also be curious if there are errors that would surface here that *wouldn't* surface when we called `signVote`? Is skipping an fsync that would fail just delaying the inevitable? It might be the case that reducing the total number of fsync calls would increase file system availability and reduce the failure rate, but we're talking about a very reasonable number of fsyncs anyway. 

I think you're right that it's not crucial to commit until we sign the vote from a "WAL-provides-double-sign-protection" but the "WAL-provides-crash-durability", but I'm not sure the right way to communicate the fact that this necessarily changes the kind of durability guarantees that the system can provide.

Reply:
Nicely put! I wonder if the `double-sign-check-height` feature could help in preventing this. We default to 0, but if we defaulted to 1 we would double check the block to see if any signatures are present with our key, if so wait till the next block.

Reply:
How critical is it though that we're able to "fast reboot"? I feel like it's only important in the case that the validator is depended on (i.e. 1/3+ vp). There is no harm to a full node outside of relying on a bit of extra i/o to get up to speed. With a validator the only harm I see is perhaps missing voting in a height/round but even that is still inconsequential. I guess we need to decide how much we want to prioritise this "fast reboot" property over performance. For me personally, I think the WAL should only provide a last defence for double-signing.

Reply:
Just checking the code. I want to double-check my logic here. It seems we flush just before sending a vote or proposal and then flush again just after (or once we read off this internalMsgQueue). Flushing beforehand is necessary so we can replay all the messages to get to the exact state we were beforehand such that we produce the same proposal or vote and the signer is okay with signing the same thing again. Is this right? And flushing afterwards is helpful insofar as we know we already produced the signed vote/proposal and don't need to ask the signer, but not critical.

Reply:
> How critical is it though that we're able to "fast reboot"? I feel like it's only important in the case that the validator is depended on (i.e. 1/3+ vp). There is no harm to a full node outside of relying on a bit of extra i/o to get up to speed. With a validator the only harm I see is perhaps missing voting in a height/round but even that is still inconsequential. I guess we need to decide how much we want to prioritise this "fast reboot" property over performance. For me personally, I think the WAL should only provide a last defence for double-signing.

This.

Reply:
Also, correct @cmwaters. We avoid an unnecessary flush in this PR.

Reply:
I agree with the general idea that "fast restarts" are relatively uncommon and the outcome of missing voting on a round is pretty inconsequential. I'm still pretty skeptical of this: 

- flushing less often means each flush to disk will be much larger, which both changes the way that it failures, but more importantly the timing of operations (e.g. the fsyncs will be bigger when they happen and take longer), which could have ripple effects throughout the system.
- I think the current state of the testing of the WAL/restart behavior is potentially petty brittle, and while I believe that this is potentially correct, we don't have tests that cover this area. 
- it doesn't feel great to make big disions about durability semantics without updating documentation etc.

Reply:
On the point of WAL/restart behavior being brittle, I do frequently see error messages like this in the e2e tests whilst replaying (i.e. directly after killing and restarting a node):
```
2021-07-13T02:37:11.0004540Z validator02    | 2021-07-13T02:36:05Z ERROR SignerServer: handleMessage err="error signing vote: step regression at height 1029 round 0. Got 2, last step 3"
2021-07-13T02:37:11.0007958Z validator02    | 2021-07-13T02:36:05Z ERROR failed signing vote err="signerEndpoint returned error #0: error signing vote: step regression at height 1029 round 0. Got 2, last step 3" height=1029 module=consensus round=0 vote={"block_id":{"hash":"","parts":{"hash":"","total":0}},"height":1029,"round":0,"signature":null,"timestamp":"2021-07-13T02:36:05.511524818Z","type":1,"validator_address":"99F30C473E41A47BC8D6675531564A73793479D2","validator_index":1}
```
Which would indicate to me that we don't always process the WAL messages correctly or if we do we nonetheless occasionally end up in states we shouldn't be in

Reply:
@tychoish what do you think could be done in terms of documentation/testing and so forth to make you more comfortable about making such a change? Do you think in the long term we should be reducing the scope (potentially removing or replacing) the WAL

Reply:
I think the WAL should be replaced or rehomed on top of our database layer (or at least implemented in terms of it in the short term,) but that has to wait until we drip support for database engines that don't provide crash-resistant semantics. (at least 0.36). The problem with ripping out the WAL entirely, is that it provides some idea of doublesign protection, and it's deeply entwined with the consensus state logic, so it can't be ripped out cleanly. 

As for docs, at the very least, this should probably be in the change log, but maybe there's additional docs that we could edit to make clear. 

For testing, I'm not sure: I think we'd definitely want to make sure that this doesn't cause regressions in the e2e tests (particularly related to pertubatons.) I also think it might be possible to construct some e2e tests with a small network and a high transaction load, and see if there's a difference in performance, and no deleterious effects? 

-----

Taking a step back from this thread. This change is likely harmless for most cases (positive and negative,) and I think it's not likely to cause a huge problem, but it does introduce some amount of risk that doesn't feel justified or mitigated, but I might be missing something.

Reply:
>  I think it's not likely to cause a huge problem, but it does introduce some amount of risk that doesn't feel justified or mitigated, but I might be missing something.

the performance considerations seem like a good enough reason to me.